// local-ComImg/src/index.js
/*
* ComImg - $1.0.5
* 功能：***
* 说明：***
* 参数： {
**  url ： required
*   param : required
* }
*/
function isArray(object){
    return object && typeof object==='object' &&
            Array == object.constructor;
}
const _uuid = () => 'WG_ComImg' + (new Date()).getTime();
class ComImg {
  constructor(option) {
    this.name = 'ComImg';
    this.version = '$1.0.5';
    this.option = option;
    this.data = {
      name: this.name,
      uuid: _uuid()
    };
    this.init();
  }
  init() {
    if(!this.option.url){
      throw new Error('url 必传');
      return;
    }
    if(!this.option.param){
      throw new Error('param 必传');
      return;
    }
  }
  getUrl(){
    return this._getTrueUrl();
  }
  _string2Url() {
    /**
     * 把String转换成URL
     * @url [Url]
     */
    if(!this.option.url){
      throw new Error('Shit');
      return;
    }
    let urlObj = {};
    let url = document.createElement('a');
    url.href = this.option.url;
    for(let key in window.location){
      urlObj[key] = url[key];
    }
    return urlObj;
  }
  _getTrueUrl() {
    /**
     * 获取转换后的url
     *
     */
    let url = this._string2Url();
    if(!url.pathname){
      throw new Error('Shit');
      return;
    }
    let pathArray = url.pathname.match(/\/[\/]?([^\/]*)\/([^!|?]*)[!]?([\S]*)/);
    if(isArray(pathArray) && pathArray.length >= 2){
      let domain = pathArray[1];
      let path = pathArray[2];
      let query = this._getTrueParam(pathArray[3]);
      if(domain.indexOf('fr-')>-1){
          return `https://${domain}.huangbaoche.com/${path}${query}`;
      }else{
        // 另外一种规则..
        let res = this._anotherRule(url.hostname);
        return `https://${res}.huangbaoche.com/${domain}/${path}${query}`;
      }
    }else{
      /**
       * 如查Array 的length 小于1 那么直接return 之前的url
       * @type {[type]}
       */
      return url.href;
    }
  }
  _anotherRule(dd){
    /**
     * [domain 返回值]
     * @dd {String} : domain
     */
    let domain = '';
    let env = ((host)=>{
      if(/hbcdn-dev/.test(host)){
        return 'dev'
      }
      if(/hbcdn-test/.test(host)){
        return 'test'
      }
      if(/hbcdn\./.test(host)){
        return 'pub'
      }
    })(dd)
    switch(env){
      case 'dev' :
        domain = 'fr-hd-dev';
        break;
      case 'test' :
        domain = 'fr-hb-t';
        break;
      case 'pub' :
        domain = 'fr-static';
        break;
      default :
        domain = 'fr-static';
    }

    return domain;
  }
  _getTrueParam(par) {
    /**
     * 获取转换后的param
     * par 这个是从URL中分析出来的，只可能是m,l,...
     * 传入的param优先级最高，传入的param会覆盖par
     */
    if(this.option.param){
      return `?${this.option.param}`;
    }else{
      return par ? `!${par}` : '';
    }

  }
}

module.exports = ComImg
